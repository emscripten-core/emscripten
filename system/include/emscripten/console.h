/*
 * Copyright 2021 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

#include <stddef.h>  // for size_t

#ifdef __cplusplus
extern "C" {
#endif

// Write directly JavaScript console.  This can be useful for debugging since it
// bypasses the stdio and filesystem sub-systems.
void emscripten_console_log(const char *utf8String __attribute__((nonnull)));
void emscripten_console_warn(const char *utf8String __attribute__((nonnull)));
void emscripten_console_error(const char *utf8String __attribute__((nonnull)));
void emscripten_console_trace(const char *utf8String __attribute__((nonnull)));

// Write to the out(), err() and dbg() JS functions directly.
// These are defined an defined in shell.js and have different behavior compared
// to console.log/err.  Under node, they write to stdout and stderr which is a
// more direct way to write output especially from worker threads.  The default
// behavior of these functions can be overridden by print and printErr, if
// provided on the Module object.  These functions are mainly intended for
// internal use.
void emscripten_out(const char *utf8String __attribute__((nonnull)));
void emscripten_err(const char *utf8String __attribute__((nonnull)));
void emscripten_dbg(const char *utf8String __attribute__((nonnull)));
void emscripten_dbg_backtrace(const char *utf8String __attribute__((nonnull)));

// Same as above but only with the length of string specified by the second
// argument.  This allows for non-NULL-terminated strings to be passed.
void emscripten_outn(const char *utf8String __attribute__((nonnull)), size_t len);
void emscripten_errn(const char *utf8String __attribute__((nonnull)), size_t len);
void emscripten_dbgn(const char *utf8String __attribute__((nonnull)), size_t len);

// Legacy/internal names for the above
#define _emscripten_out(x) emscripten_out(x)
#define _emscripten_err(x) emscripten_err(x)
#define _emscripten_dbg(x) emscripten_dbg(x)

// Similar to the above functions but operate with printf-like semantics.
void emscripten_console_logf(const char *format __attribute__((nonnull)), ...) __attribute__((__format__(printf, 1, 2)));
void emscripten_console_warnf(const char *format __attribute__((nonnull)), ...) __attribute__((__format__(printf, 1, 2)));
void emscripten_console_errorf(const char *format __attribute__((nonnull)), ...)__attribute__((__format__(printf, 1, 2)));
void emscripten_console_tracef(const char *format __attribute__((nonnull)), ...)__attribute__((__format__(printf, 1, 2)));
void emscripten_outf(const char *format __attribute__((nonnull)), ...) __attribute__((__format__(printf, 1, 2)));
void emscripten_errf(const char *format __attribute__((nonnull)), ...) __attribute__((__format__(printf, 1, 2)));
void emscripten_dbgf(const char *format __attribute__((nonnull)), ...) __attribute__((__format__(printf, 1, 2)));
void emscripten_dbg_backtracef(const char *format __attribute__((nonnull)), ...) __attribute__((__format__(printf, 1, 2)));

// Legacy/internal names for the above
#define _emscripten_outf(format, ...) emscripten_outf(format, ##__VA_ARGS__)
#define _emscripten_errf(format, ...) emscripten_errf(format, ##__VA_ARGS__)
#define _emscripten_dbgf(format, ...) emscripten_dbgf(format, ##__VA_ARGS__)

#ifdef __cplusplus
}
#endif
