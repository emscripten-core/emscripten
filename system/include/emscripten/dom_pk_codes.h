/*
 * Copyright 2018 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 *
 * This file was automatically generated from script
 * tools/maint/create_dom_pk_codes.py. Edit that file to make changes here.
 * Then run:
 *
 *   tools/maint/create_dom_pk_codes.py
 *
 * in Emscripten root directory to regenerate this file.
 */

#pragma once

#define DOM_PK_CODE_TYPE int

#define DOM_PK_UNKNOWN              0x0000 /* "Unidentified"       */
#define DOM_PK_ESCAPE               0x0001 /* "Escape"             */
#define DOM_PK_0                    0x0002 /* "Digit0"             */
#define DOM_PK_1                    0x0003 /* "Digit1"             */
#define DOM_PK_2                    0x0004 /* "Digit2"             */
#define DOM_PK_3                    0x0005 /* "Digit3"             */
#define DOM_PK_4                    0x0006 /* "Digit4"             */
#define DOM_PK_5                    0x0007 /* "Digit5"             */
#define DOM_PK_6                    0x0008 /* "Digit6"             */
#define DOM_PK_7                    0x0009 /* "Digit7"             */
#define DOM_PK_8                    0x000A /* "Digit8"             */
#define DOM_PK_9                    0x000B /* "Digit9"             */
#define DOM_PK_MINUS                0x000C /* "Minus"              */
#define DOM_PK_EQUAL                0x000D /* "Equal"              */
#define DOM_PK_BACKSPACE            0x000E /* "Backspace"          */
#define DOM_PK_TAB                  0x000F /* "Tab"                */
#define DOM_PK_Q                    0x0010 /* "KeyQ"               */
#define DOM_PK_W                    0x0011 /* "KeyW"               */
#define DOM_PK_E                    0x0012 /* "KeyE"               */
#define DOM_PK_R                    0x0013 /* "KeyR"               */
#define DOM_PK_T                    0x0014 /* "KeyT"               */
#define DOM_PK_Y                    0x0015 /* "KeyY"               */
#define DOM_PK_U                    0x0016 /* "KeyU"               */
#define DOM_PK_I                    0x0017 /* "KeyI"               */
#define DOM_PK_O                    0x0018 /* "KeyO"               */
#define DOM_PK_P                    0x0019 /* "KeyP"               */
#define DOM_PK_BRACKET_LEFT         0x001A /* "BracketLeft"        */
#define DOM_PK_BRACKET_RIGHT        0x001B /* "BracketRight"       */
#define DOM_PK_ENTER                0x001C /* "Enter"              */
#define DOM_PK_CONTROL_LEFT         0x001D /* "ControlLeft"        */
#define DOM_PK_A                    0x001E /* "KeyA"               */
#define DOM_PK_S                    0x001F /* "KeyS"               */
#define DOM_PK_D                    0x0020 /* "KeyD"               */
#define DOM_PK_F                    0x0021 /* "KeyF"               */
#define DOM_PK_G                    0x0022 /* "KeyG"               */
#define DOM_PK_H                    0x0023 /* "KeyH"               */
#define DOM_PK_J                    0x0024 /* "KeyJ"               */
#define DOM_PK_K                    0x0025 /* "KeyK"               */
#define DOM_PK_L                    0x0026 /* "KeyL"               */
#define DOM_PK_SEMICOLON            0x0027 /* "Semicolon"          */
#define DOM_PK_QUOTE                0x0028 /* "Quote"              */
#define DOM_PK_BACKQUOTE            0x0029 /* "Backquote"          */
#define DOM_PK_SHIFT_LEFT           0x002A /* "ShiftLeft"          */
#define DOM_PK_BACKSLASH            0x002B /* "Backslash"          */
#define DOM_PK_Z                    0x002C /* "KeyZ"               */
#define DOM_PK_X                    0x002D /* "KeyX"               */
#define DOM_PK_C                    0x002E /* "KeyC"               */
#define DOM_PK_V                    0x002F /* "KeyV"               */
#define DOM_PK_B                    0x0030 /* "KeyB"               */
#define DOM_PK_N                    0x0031 /* "KeyN"               */
#define DOM_PK_M                    0x0032 /* "KeyM"               */
#define DOM_PK_COMMA                0x0033 /* "Comma"              */
#define DOM_PK_PERIOD               0x0034 /* "Period"             */
#define DOM_PK_SLASH                0x0035 /* "Slash"              */
#define DOM_PK_SHIFT_RIGHT          0x0036 /* "ShiftRight"         */
#define DOM_PK_NUMPAD_MULTIPLY      0x0037 /* "NumpadMultiply"     */
#define DOM_PK_ALT_LEFT             0x0038 /* "AltLeft"            */
#define DOM_PK_SPACE                0x0039 /* "Space"              */
#define DOM_PK_CAPS_LOCK            0x003A /* "CapsLock"           */
#define DOM_PK_F1                   0x003B /* "F1"                 */
#define DOM_PK_F2                   0x003C /* "F2"                 */
#define DOM_PK_F3                   0x003D /* "F3"                 */
#define DOM_PK_F4                   0x003E /* "F4"                 */
#define DOM_PK_F5                   0x003F /* "F5"                 */
#define DOM_PK_F6                   0x0040 /* "F6"                 */
#define DOM_PK_F7                   0x0041 /* "F7"                 */
#define DOM_PK_F8                   0x0042 /* "F8"                 */
#define DOM_PK_F9                   0x0043 /* "F9"                 */
#define DOM_PK_F10                  0x0044 /* "F10"                */
#define DOM_PK_PAUSE                0x0045 /* "Pause"              */
#define DOM_PK_SCROLL_LOCK          0x0046 /* "ScrollLock"         */
#define DOM_PK_NUMPAD_7             0x0047 /* "Numpad7"            */
#define DOM_PK_NUMPAD_8             0x0048 /* "Numpad8"            */
#define DOM_PK_NUMPAD_9             0x0049 /* "Numpad9"            */
#define DOM_PK_NUMPAD_SUBTRACT      0x004A /* "NumpadSubtract"     */
#define DOM_PK_NUMPAD_4             0x004B /* "Numpad4"            */
#define DOM_PK_NUMPAD_5             0x004C /* "Numpad5"            */
#define DOM_PK_NUMPAD_6             0x004D /* "Numpad6"            */
#define DOM_PK_NUMPAD_ADD           0x004E /* "NumpadAdd"          */
#define DOM_PK_NUMPAD_1             0x004F /* "Numpad1"            */
#define DOM_PK_NUMPAD_2             0x0050 /* "Numpad2"            */
#define DOM_PK_NUMPAD_3             0x0051 /* "Numpad3"            */
#define DOM_PK_NUMPAD_0             0x0052 /* "Numpad0"            */
#define DOM_PK_NUMPAD_DECIMAL       0x0053 /* "NumpadDecimal"      */
#define DOM_PK_PRINT_SCREEN         0x0054 /* "PrintScreen"        */
#define DOM_PK_INTL_BACKSLASH       0x0056 /* "IntlBackslash"      */
#define DOM_PK_F11                  0x0057 /* "F11"                */
#define DOM_PK_F12                  0x0058 /* "F12"                */
#define DOM_PK_NUMPAD_EQUAL         0x0059 /* "NumpadEqual"        */
#define DOM_PK_F13                  0x0064 /* "F13"                */
#define DOM_PK_F14                  0x0065 /* "F14"                */
#define DOM_PK_F15                  0x0066 /* "F15"                */
#define DOM_PK_F16                  0x0067 /* "F16"                */
#define DOM_PK_F17                  0x0068 /* "F17"                */
#define DOM_PK_F18                  0x0069 /* "F18"                */
#define DOM_PK_F19                  0x006A /* "F19"                */
#define DOM_PK_F20                  0x006B /* "F20"                */
#define DOM_PK_F21                  0x006C /* "F21"                */
#define DOM_PK_F22                  0x006D /* "F22"                */
#define DOM_PK_F23                  0x006E /* "F23"                */
#define DOM_PK_KANA_MODE            0x0070 /* "KanaMode"           */
#define DOM_PK_LANG_2               0x0071 /* "Lang2"              */
#define DOM_PK_LANG_1               0x0072 /* "Lang1"              */
#define DOM_PK_INTL_RO              0x0073 /* "IntlRo"             */
#define DOM_PK_F24                  0x0076 /* "F24"                */
#define DOM_PK_CONVERT              0x0079 /* "Convert"            */
#define DOM_PK_NON_CONVERT          0x007B /* "NonConvert"         */
#define DOM_PK_INTL_YEN             0x007D /* "IntlYen"            */
#define DOM_PK_NUMPAD_COMMA         0x007E /* "NumpadComma"        */
#define DOM_PK_PASTE                0xE00A /* "Paste"              */
#define DOM_PK_MEDIA_TRACK_PREVIOUS 0xE010 /* "MediaTrackPrevious" */
#define DOM_PK_CUT                  0xE017 /* "Cut"                */
#define DOM_PK_COPY                 0xE018 /* "Copy"               */
#define DOM_PK_MEDIA_TRACK_NEXT     0xE019 /* "MediaTrackNext"     */
#define DOM_PK_NUMPAD_ENTER         0xE01C /* "NumpadEnter"        */
#define DOM_PK_CONTROL_RIGHT        0xE01D /* "ControlRight"       */
#define DOM_PK_AUDIO_VOLUME_MUTE    0xE020 /* "AudioVolumeMute"    */
#define DOM_PK_AUDIO_VOLUME_MUTE    0xE020 /* "VolumeMute"         */
#define DOM_PK_LAUNCH_APP_2         0xE021 /* "LaunchApp2"         */
#define DOM_PK_MEDIA_PLAY_PAUSE     0xE022 /* "MediaPlayPause"     */
#define DOM_PK_MEDIA_STOP           0xE024 /* "MediaStop"          */
#define DOM_PK_EJECT                0xE02C /* "Eject"              */
#define DOM_PK_AUDIO_VOLUME_DOWN    0xE02E /* "AudioVolumeDown"    */
#define DOM_PK_AUDIO_VOLUME_DOWN    0xE02E /* "VolumeDown"         */
#define DOM_PK_AUDIO_VOLUME_UP      0xE030 /* "AudioVolumeUp"      */
#define DOM_PK_AUDIO_VOLUME_UP      0xE030 /* "VolumeUp"           */
#define DOM_PK_BROWSER_HOME         0xE032 /* "BrowserHome"        */
#define DOM_PK_NUMPAD_DIVIDE        0xE035 /* "NumpadDivide"       */
#define DOM_PK_ALT_RIGHT            0xE038 /* "AltRight"           */
#define DOM_PK_HELP                 0xE03B /* "Help"               */
#define DOM_PK_NUM_LOCK             0xE045 /* "NumLock"            */
#define DOM_PK_HOME                 0xE047 /* "Home"               */
#define DOM_PK_ARROW_UP             0xE048 /* "ArrowUp"            */
#define DOM_PK_PAGE_UP              0xE049 /* "PageUp"             */
#define DOM_PK_ARROW_LEFT           0xE04B /* "ArrowLeft"          */
#define DOM_PK_ARROW_RIGHT          0xE04D /* "ArrowRight"         */
#define DOM_PK_END                  0xE04F /* "End"                */
#define DOM_PK_ARROW_DOWN           0xE050 /* "ArrowDown"          */
#define DOM_PK_PAGE_DOWN            0xE051 /* "PageDown"           */
#define DOM_PK_INSERT               0xE052 /* "Insert"             */
#define DOM_PK_DELETE               0xE053 /* "Delete"             */
#define DOM_PK_META_LEFT            0xE05B /* "MetaLeft"           */
#define DOM_PK_OS_LEFT              0xE05B /* "OSLeft"             */
#define DOM_PK_META_RIGHT           0xE05C /* "MetaRight"          */
#define DOM_PK_OS_RIGHT             0xE05C /* "OSRight"            */
#define DOM_PK_CONTEXT_MENU         0xE05D /* "ContextMenu"        */
#define DOM_PK_POWER                0xE05E /* "Power"              */
#define DOM_PK_BROWSER_SEARCH       0xE065 /* "BrowserSearch"      */
#define DOM_PK_BROWSER_FAVORITES    0xE066 /* "BrowserFavorites"   */
#define DOM_PK_BROWSER_REFRESH      0xE067 /* "BrowserRefresh"     */
#define DOM_PK_BROWSER_STOP         0xE068 /* "BrowserStop"        */
#define DOM_PK_BROWSER_FORWARD      0xE069 /* "BrowserForward"     */
#define DOM_PK_BROWSER_BACK         0xE06A /* "BrowserBack"        */
#define DOM_PK_LAUNCH_APP_1         0xE06B /* "LaunchApp1"         */
#define DOM_PK_LAUNCH_MAIL          0xE06C /* "LaunchMail"         */
#define DOM_PK_LAUNCH_MEDIA_PLAYER  0xE06D /* "LaunchMediaPlayer"  */
#define DOM_PK_MEDIA_SELECT         0xE06D /* "MediaSelect"        */

#ifdef __cplusplus
extern "C" {
#endif
/* Maps the EmscriptenKeyboardEvent::code field from emscripten/html5.h to one of the DOM_PK codes above. */
DOM_PK_CODE_TYPE emscripten_compute_dom_pk_code(const char *keyCodeString);

/* Returns the string representation of the given key code ID. Useful for debug printing. */
const char *emscripten_dom_pk_code_to_string(DOM_PK_CODE_TYPE code);
#ifdef __cplusplus
}
#endif
