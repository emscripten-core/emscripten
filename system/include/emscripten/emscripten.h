/*
 * Copyright 2012 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

#pragma once

/**
 * This file contains a few useful things for compiling C/C++ code
 * with Emscripten.
 *
 * Documentation for the public APIs defined in this file must be updated in: 
 *    site/source/docs/api_reference/emscripten.h.rst
 * A prebuilt local version of the documentation is available at: 
 *    site/build/text/docs/api_reference/emscripten.h.txt
 * You can also build docs locally as HTML or other formats in site/
 * An online HTML version (which may be of a different version of Emscripten)
 *    is up at http://kripken.github.io/emscripten-site/docs/api_reference/emscripten.h.html
 */

#include "em_asm.h"
#include "em_macros.h"
#include "em_types.h"
#include "em_js.h"
#include "promise.h"
#include "wget.h"
#include "version.h"

#ifdef __cplusplus
extern "C" {
#endif

void emscripten_run_script(const char *script __attribute__((nonnull)));
int emscripten_run_script_int(const char *script __attribute__((nonnull)));
char *emscripten_run_script_string(const char *script __attribute__((nonnull)));
void emscripten_async_run_script(const char *script __attribute__((nonnull)), int millis);
void emscripten_async_load_script(const char *script __attribute__((nonnull)), em_callback_func onload, em_callback_func onerror);

void emscripten_set_main_loop(em_callback_func func, int fps, bool simulate_infinite_loop);

#define EM_TIMING_SETTIMEOUT 0
#define EM_TIMING_RAF 1
#define EM_TIMING_SETIMMEDIATE 2

int emscripten_set_main_loop_timing(int mode, int value);
void emscripten_get_main_loop_timing(int *mode, int *value); // Pass a null pointer to skip receiving that particular value
void emscripten_set_main_loop_arg(em_arg_callback_func func, void *arg, int fps, bool simulate_infinite_loop);
void emscripten_pause_main_loop(void);
void emscripten_resume_main_loop(void);
void emscripten_cancel_main_loop(void);

typedef void (*em_socket_callback)(int fd, void *userData);
typedef void (*em_socket_error_callback)(int fd, int err, const char* msg, void *userData);

void emscripten_set_socket_error_callback(void *userData, em_socket_error_callback callback);
void emscripten_set_socket_open_callback(void *userData, em_socket_callback callback);
void emscripten_set_socket_listen_callback(void *userData, em_socket_callback callback);
void emscripten_set_socket_connection_callback(void *userData, em_socket_callback callback);
void emscripten_set_socket_message_callback(void *userData, em_socket_callback callback);
void emscripten_set_socket_close_callback(void *userData, em_socket_callback callback);

void _emscripten_push_main_loop_blocker(em_arg_callback_func func, void *arg, const char *name);
void _emscripten_push_uncounted_main_loop_blocker(em_arg_callback_func func, void *arg, const char *name);
#define emscripten_push_main_loop_blocker(func, arg) \
  _emscripten_push_main_loop_blocker(func, arg, #func)
#define emscripten_push_uncounted_main_loop_blocker(func, arg) \
  _emscripten_push_uncounted_main_loop_blocker(func, arg, #func)

void emscripten_set_main_loop_expected_blockers(int num);

void emscripten_async_call(em_arg_callback_func func, void *arg, int millis);

void emscripten_exit_with_live_runtime(void) __attribute__((__noreturn__));
void emscripten_force_exit(int status) __attribute__((__noreturn__));

double emscripten_get_device_pixel_ratio(void);

char *emscripten_get_window_title(void);
void emscripten_set_window_title(const char *);
void emscripten_get_screen_size(int *width __attribute__((nonnull)), int *height __attribute__((nonnull)));
void emscripten_hide_mouse(void);
void emscripten_set_canvas_size(int width, int height) __attribute__((deprecated("This variant does not allow specifying the target canvas", "Use emscripten_set_canvas_element_size() instead")));
void emscripten_get_canvas_size(int *width __attribute__((nonnull)), int *height __attribute__((nonnull)), int *isFullscreen __attribute__((nonnull))) __attribute__((deprecated("This variant does not allow specifying the target canvas", "Use emscripten_get_canvas_element_size() and emscripten_get_fullscreen_status() instead")));

double emscripten_get_now(void);
float emscripten_random(void);

// IDB

typedef void (*em_idb_onload_func)(void*, void*, int);
void emscripten_idb_async_load(const char *db_name __attribute__((nonnull)), const char *file_id __attribute__((nonnull)), void* arg, em_idb_onload_func onload, em_arg_callback_func onerror);
void emscripten_idb_async_store(const char *db_name __attribute__((nonnull)), const char *file_id __attribute__((nonnull)), void* ptr, int num, void* arg, em_arg_callback_func onstore, em_arg_callback_func onerror);
void emscripten_idb_async_delete(const char *db_name __attribute__((nonnull)), const char *file_id __attribute__((nonnull)), void* arg, em_arg_callback_func ondelete, em_arg_callback_func onerror);
typedef void (*em_idb_exists_func)(void*, int);
void emscripten_idb_async_exists(const char *db_name __attribute__((nonnull)), const char *file_id __attribute__((nonnull)), void* arg, em_idb_exists_func oncheck, em_arg_callback_func onerror);
void emscripten_idb_async_clear(const char *db_name __attribute__((nonnull)), void* arg, em_arg_callback_func onclear, em_arg_callback_func onerror);

// IDB "sync"

void emscripten_idb_load(const char *db_name, const char *file_id, void** pbuffer, int* pnum, int *perror);
void emscripten_idb_store(const char *db_name, const char *file_id, void* buffer, int num, int *perror);
void emscripten_idb_delete(const char *db_name, const char *file_id, int *perror);
void emscripten_idb_exists(const char *db_name, const char *file_id, int* pexists, int *perror);
void emscripten_idb_clear(const char *db_name, int *perror);

void emscripten_idb_load_blob(const char *db_name, const char *file_id, int* pblob, int *perror);
void emscripten_idb_store_blob(const char *db_name, const char *file_id, void* buffer, int num, int *perror);
void emscripten_idb_read_from_blob(int blob, int start, int num, void* buffer);
void emscripten_idb_free_blob(int blob);

// other async utilities

int emscripten_run_preload_plugins(const char* file, em_str_callback_func onload, em_str_callback_func onerror);

typedef void (*em_run_preload_plugins_data_onload_func)(void*, const char*);
void emscripten_run_preload_plugins_data(char* data, int size, const char *suffix, void *arg, em_run_preload_plugins_data_onload_func onload, em_arg_callback_func onerror);

void emscripten_lazy_load_code(void);

// show an error on some renamed methods
#define emscripten_async_prepare(...) _Pragma("GCC error(\"emscripten_async_prepare has been replaced by emscripten_run_preload_plugins\")")
#define emscripten_async_prepare_data(...) _Pragma("GCC error(\"emscripten_async_prepare_data has been replaced by emscripten_run_preload_plugins_data\")")

// worker APIs

typedef int worker_handle;

worker_handle emscripten_create_worker(const char *url);
void emscripten_destroy_worker(worker_handle worker);

typedef void (*em_worker_callback_func)(char*, int, void*);
void emscripten_call_worker(worker_handle worker, const char *funcname, char *data, int size, em_worker_callback_func callback, void *arg);
void emscripten_worker_respond(char *data, int size);
void emscripten_worker_respond_provisionally(char *data, int size);

int emscripten_get_worker_queue_size(worker_handle worker);

// misc.

long emscripten_get_compiler_setting(const char *name);
int emscripten_has_asyncify(void);

void emscripten_debugger(void);

// Forward declare FILE from musl libc headers to avoid needing to #include <stdio.h> from emscripten.h
struct _IO_FILE;
typedef struct _IO_FILE FILE;

char *emscripten_get_preloaded_image_data(const char *path, int *w, int *h);
char *emscripten_get_preloaded_image_data_from_FILE(FILE *file, int *w, int *h);

#define EM_LOG_CONSOLE   1
#define EM_LOG_WARN      2
#define EM_LOG_ERROR     4
#define EM_LOG_C_STACK   8
#define EM_LOG_JS_STACK 16
#define EM_LOG_DEMANGLE 32  // deprecated
#pragma clang deprecated(EM_LOG_DEMANGLE)
#define EM_LOG_NO_PATHS 64
#define EM_LOG_FUNC_PARAMS 128  // deprecated
#pragma clang deprecated(EM_LOG_FUNC_PARAMS)
#define EM_LOG_DEBUG    256
#define EM_LOG_INFO     512

void emscripten_log(int flags, const char* format, ...);

int emscripten_get_callstack(int flags, char *out, int maxbytes);

int emscripten_print_double(double x, char *to, signed max);

typedef void (*em_scan_func)(void*, void*);
void emscripten_scan_registers(em_scan_func func);
void emscripten_scan_stack(em_scan_func func);

// Asynchronous version of dlopen.  Since WebAssembly module loading in general
// is asynchronous the normal dlopen function can't be used in all situations.
typedef void (*em_dlopen_callback)(void* user_data, void* handle);
void emscripten_dlopen(const char *filename, int flags, void* user_data, em_dlopen_callback onsuccess, em_arg_callback_func onerror);

// Promisified version of emscripten_dlopen
// The returned promise will resolve once the dso has been loaded.  Its up to
// the caller to call emscripten_promise_destroy on this promise.
em_promise_t emscripten_dlopen_promise(const char *filename, int flags);

void emscripten_throw_number(double number);
void emscripten_throw_string(const char *utf8String);

/* ===================================== */
/* Internal APIs. Be careful with these. */
/* ===================================== */

void emscripten_sleep(unsigned int ms);

#ifdef __cplusplus
}
#endif
