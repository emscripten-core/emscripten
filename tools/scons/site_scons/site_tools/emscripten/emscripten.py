#!/usr/bin/env python2
# Copyright 2012 The Emscripten Authors.  All rights reserved.
# Emscripten is available under two separate licenses, the MIT license and the
# University of Illinois/NCSA Open Source License.  Both these licenses can be
# found in the LICENSE file.

from __future__ import print_function
import os


def path_to_bin(emscripten_root, bin_name):
  if emscripten_root is None:
    return bin_name
  else:
    return os.path.join(emscripten_root, bin_name)


def generate(env, **kw):
  """ SCons tool entry point """

  # SCons does not by default invoke the compiler with the
  # environment variables from the parent calling process,
  # so manually route all environment variables referenced
  # by Emscripten to the child.
  emscripten_env_vars = ['EM_CACHE', 'EMCC_DEBUG', 'EM_CONFIG',
    'EMMAKEN_JUST_CONFIGURE', 'EMCC_CFLAGS', 'EMCC_TEMP_DIR',
    'EMCC_AUTODEBUG', 'EM_COMPILER_WRAPPER', 'MOZ_DISABLE_AUTO_SAFE_MODE',
    'EMCC_STDERR_FILE', 'EMSCRIPTEN_SUPPRESS_USAGE_WARNING', 'NODE_PATH',
    'EMCC_JSOPT_MIN_CHUNK_SIZE', 'EMCC_JSOPT_MAX_CHUNK_SIZE',
    'EMCC_CORES', 'EMCC_NO_OPT_SORT', 'EMCC_BUILD_DIR',
    'EMCC_DEBUG_SAVE', 'EMCC_SKIP_SANITY_CHECK', 'EM_PKG_CONFIG_PATH',
    'EMCC_CLOSURE_ARGS', 'EMCC_FORCE_STDLIBS',
    'EMCC_ONLY_FORCED_STDLIBS', 'EM_PORTS', 'IDL_CHECKS', 'IDL_VERBOSE']

  pkg_config_vars = {
    'EMSCONS_PKG_CONFIG_LIBDIR': 'PKG_CONFIG_LIBDIR',
    'EMSCONS_PKG_CONFIG_PATH': 'PKG_CONFIG_PATH',
  }

  for var in emscripten_env_vars:
    if var in os.environ:
      env['ENV'][var] = os.environ[var]

  for var in pkg_config_vars:
    if var in os.environ:
      real_key = pkg_config_vars[var]
      env['ENV'][real_key] = os.environ[var]

  # Binary paths will be constructed from here if available.
  # Otherwise they are assumed to be in the PATH.
  emscripten_root = os.environ.get('EMSCRIPTEN_ROOT')

  env.Replace(CC=path_to_bin(emscripten_root, 'emcc'))
  env.Replace(CXX=path_to_bin(emscripten_root, 'em++'))
  # LINK uses smark_link by default which will choose
  # either emcc or em++ depending on if there are any C++ sources
  # in the program, so no need to change that.
  # SHLINK and LDMODULE should use LINK so no
  # need to change them here

  env.Replace(AR=path_to_bin(emscripten_root, 'emar'))
  env.Replace(RANLIB=path_to_bin(emscripten_root, 'emranlib'))

  env.Replace(PROGSUFFIX='.js')


def exists(env):
  """ NOOP method required by SCons """
  return 1
