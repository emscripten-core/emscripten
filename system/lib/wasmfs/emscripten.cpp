// Copyright 2023 The Emscripten Authors.  All rights reserved.
// Emscripten is available under two separate licenses, the MIT license and the
// University of Illinois/NCSA Open Source License.  Both these licenses can be
// found in the LICENSE file.

//
// This file contains implementations of emscripten APIs that are compatible
// with WasmFS. These replace APIs in src/library*js, and basically do things
// in a simpler manner for the situation where the FS is in wasm and not JS
// (in particular, these implementations avoid calling from JS to wasm, and
// dependency issues that arise from that).
//

#include <emscripten.h>
#include <stdio.h>

#include <string>

#include "file.h"
#include "file_table.h"
#include "wasmfs.h"

namespace wasmfs {

// Given an fd, returns the string path that the fd refers to.
// TODO: full error handling
// TODO: maybe make this a public API, as it is useful for debugging
std::string getPath(int fd) {
  auto fileTable = wasmfs::wasmFS.getFileTable().locked();

  auto openFile = fileTable.getEntry(fd);
  if (!openFile) {
    return "!";
  }

  auto curr = openFile->locked().getFile();
  std::string result = "";
  while (curr != wasmfs::wasmFS.getRootDirectory()) {
    auto parent = curr->locked().getParent();
    // Check if the parent exists. The parent may not exist if curr was
    // unlinked.
    if (!parent) {
      return "?/" + result;
    }

    auto parentDir = parent->dynCast<wasmfs::Directory>();
    auto name = parentDir->locked().getName(curr);
    result = '/' + name + result;
    curr = parentDir;
  }
  return result;
}

} // namespace wasmfs

extern "C" {

char* emscripten_get_preloaded_image_data_from_FILE(FILE* file,
                                                    int* w,
                                                    int* h) {
  auto fd = fileno(file);
  if (fd < 0) {
    return 0;
  }

  auto path = wasmfs::getPath(fd);
  return emscripten_get_preloaded_image_data(path.c_str(), w, h);
}

} // extern "C"
