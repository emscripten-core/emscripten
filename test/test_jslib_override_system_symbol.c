#include <emscripten.h>
#include <emscripten/html5.h>
#include <assert.h>
#include <GLES3/gl3.h>
#include <stdio.h>

GLuint what_got_created(void);

int main() {
  EmscriptenWebGLContextAttributes attrs;
  emscripten_webgl_init_context_attributes(&attrs);
  attrs.majorVersion = 2;
  emscripten_webgl_make_context_current(emscripten_webgl_create_context("canvas", &attrs));

  GLuint createType = GL_UNSIGNED_BYTE;
  glTexImage3D(GL_TEXTURE_3D, 0, GL_RGBA, 1, 1, 1, 0, GL_RGBA, createType, 0);
  GLuint whatGotCreated = what_got_created();
  printf("Created texture of type 0x%x\n", whatGotCreated);
  assert(createType == whatGotCreated);
  return 0;
}
