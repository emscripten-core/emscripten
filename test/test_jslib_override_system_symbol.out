Created texture of type 0x1401
